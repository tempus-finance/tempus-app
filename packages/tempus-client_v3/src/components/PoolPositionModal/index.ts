import PoolPositionModal from './PoolPositionModal';

export default PoolPositionModal;
