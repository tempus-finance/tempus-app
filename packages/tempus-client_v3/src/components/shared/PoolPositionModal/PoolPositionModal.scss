.tc__poolPositionModal-info {
  margin-top: 24px;
  width: 736px;
}

.tc__poolPositionModal-chart {
  margin-top: 16px;
  width: 736px;
  height: 345px;
}

.tc__poolPositionModal-actions {
  width: 100%;
  display: flex;
  justify-content: center;
}
