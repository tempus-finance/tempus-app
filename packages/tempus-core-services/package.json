{
  "name": "tempus-core-services",
  "version": "0.0.1",
  "private": true,
  "description": "Tempus core Services",
  "keywords": [
    "typescript",
    "boilerplate",
    "esbuild"
  ],
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "scripts": {
    "lint": "eslint src/ --ext .js,.jsx,.ts,.tsx",
    "test": "jest",
    "clean": "rm -rf dist build package",
    "ts-node": "ts-node",
    "docs": "typedoc --entryPoints src/index.ts",
    "build": "npm run clean && tsc -p tsconfig.json",
    "build-all": "npm run clean && npm run build && npm run esbuild-node && npm run esbuild-node",
    "esbuild-node": "esbuild src/index.ts --bundle --platform=node --minify --sourcemap=external --outfile=dist/esbuild/index.js",
    "esbuild-node:dev": "esbuild src/index.ts --bundle --platform=node --sourcemap=external --outfile=dist/esbuild/index.js",
    "esbuild-node:watch": "esbuild src/index.ts --bundle --platform=node --watch --sourcemap=external --outfile=dist/esbuild/index.js"
  },
  "dependencies": {
    "@ethersproject/providers": "^5.5.2",
    "@types/jest": "^24.9.1",
    "@types/node": "^12.20.15",
    "axios": "^0.22.0",
    "date-fns": "^2.28.0",
    "ethers": "^5.4.6"
  },
  "devDependencies": {
    "esbuild": "^0.11.11",
    "eslint": "^7.22.0",
    "jest": "^26.6.3",
    "ts-jest": "^26.5.4",
    "ts-node": "^9.1.1",
    "typedoc": "^0.20.35"
  },
  "peerDependencies": {
    "@typescript-eslint/eslint-plugin": "^4.28.1",
    "@typescript-eslint/parser": "^4.28.1",
    "eslint-config-airbnb-typescript": "^12.3.1"
  }
}
