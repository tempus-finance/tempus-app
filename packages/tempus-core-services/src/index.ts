import ERC20ABI from './abi/ERC20.json';
import VaultABI from './abi/Vault.json';

export { ERC20ABI, VaultABI };

export type { ERC20 } from './abi';

export type {
  AvailableToDeposit,
  Chain,
  ChainConfig,
  Config,
  DynamicTempusPool,
  PoolShares,
  ProtocolDisplayName,
  ProtocolName,
  Ticker,
  NativeTokenTicker,
  TempusPool,
  TokenPrecision,
  TokenTypePrecision,
  YearnData,
  YearnDataApy,
  YearnDataToken,
} from './interfaces';

export type {
  DepositedEvent,
  ERC20TokenService,
  PoolBalanceChangedEvent,
  PoolBalanceChangedEventListener,
  RedeemedEvent,
  TempusAMMJoinKind,
  TempusAMMExitKind,
  TempusAMMService,
  TempusControllerService,
  TempusPoolService,
  TransferEventListener,
  VaultService,
  StatisticsService,
  StorageService,
  SwapEvent,
} from './services';

export type { Numberish } from './datastructures';

export { Decimal, ZERO, ONE, DEFAULT_DECIMAL_PRECISION } from './datastructures';

export { SwapKind } from './services';

export {
  getChainlinkFeed,
  getCoingeckoRate,
  getDefaultProvider,
  getERC20TokenService,
  initServices,
  getServices,
  getDefinedServices,
} from './services';

export {
  NumberUtils,
  DecimalUtils,
  capitalize,
  chainIdToChainName,
  chainIdHexToChainName,
  chainNameToHexChainId,
  chainToTicker,
  decreasePrecision,
  div18f,
  getCookie,
  getLibrary,
  getPastDaysNumber,
  getProvider,
  getProviderFromSignerOrProvider,
  getRangeFrom,
  getTokenPrecision,
  increasePrecision,
  isZeroString,
  mul18f,
  prettifyChainName,
  prettifyChainNameLong,
  rayToDai,
  reverseString,
  shortenAccount,
  wadToDai,
  wait,
  prettifyProtocolName,
} from './utils';

export * from './constants';
