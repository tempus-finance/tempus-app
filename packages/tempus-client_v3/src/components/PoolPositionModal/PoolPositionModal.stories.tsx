import { ComponentStory, ComponentMeta } from '@storybook/react';
import { Decimal } from 'tempus-core-services';
import { PoolPositionModal } from './PoolPositionModal';

export default {
  title: 'PoolPositionModal',
  component: PoolPositionModal,
  argTypes: {
    term: {
      control: {
        type: 'date',
      },
    },
  },
} as ComponentMeta<typeof PoolPositionModal>;

const Template: ComponentStory<typeof PoolPositionModal> = args => <PoolPositionModal {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  apr: 0.1,
  term: new Date(Date.UTC(2022, 2, 4)),
  profitLoss: new Decimal(10.29651),
  balance: new Decimal(20.9294),
  totalYieldEarned: new Decimal(1.12345),
  projectedTotalYield: new Decimal(3.181215),
  tokenExchangeRate: new Decimal(2000),
  tokenDecimals: 4,
  tokenTicker: 'ETH',
  chartData: [
    {
      x: new Date(Date.UTC(2022, 4, 20)),
      y: 1,
    },
    {
      x: new Date(Date.UTC(2022, 4, 21)),
      y: 4,
    },
    {
      x: new Date(Date.UTC(2022, 4, 22)),
      y: 3.4,
    },
    {
      x: new Date(Date.UTC(2022, 4, 23)),
      y: 5,
    },
  ],
};
