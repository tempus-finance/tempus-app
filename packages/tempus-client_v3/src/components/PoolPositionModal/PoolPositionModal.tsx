import { FC, useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import { Decimal, Ticker } from 'tempus-core-services';
import { ChartDataPoint, ChartDot, DateChart, ChartDotVariant } from '../shared/Chart';
import ActionButton from '../shared/ActionButton';
import Modal from '../shared/Modal';
import PoolPositionCard from '../shared/PoolPositionCard';

import './PoolPositionModal.scss';

interface PoolPositionModalProps {
  apr: number;
  term: Date;
  profitLoss: Decimal;
  balance: Decimal;
  totalYieldEarned: Decimal;
  projectedTotalYield: Decimal;
  tokenDecimals: number;
  tokenExchangeRate: Decimal;
  tokenTicker: Ticker;
  chartData: ChartDataPoint<Date, number>[];
  onWithdraw: () => void;
}

const PoolPositionModal: FC<PoolPositionModalProps> = props => {
  const {
    apr,
    term,
    profitLoss,
    balance,
    totalYieldEarned,
    projectedTotalYield,
    tokenDecimals,
    tokenExchangeRate,
    tokenTicker,
    chartData,
    onWithdraw,
  } = props;

  const { t } = useTranslation();

  const chartDot = useCallback(
    (_x, _y, index, cx, cy) => {
      let variant: ChartDotVariant = 'plus';
      if (index > 1) {
        const currentValue = chartData[index].y;
        const previousValue = chartData[index - 1].y;

        if (currentValue < previousValue) {
          variant = 'minus';
        }
      }

      return <ChartDot variant={variant} centerX={cx} centerY={cy} key={`chart-dot-${cx}-${cy}`} />;
    },
    [chartData],
  );

  return (
    <Modal open onClose={() => {}} title={t('PoolPositionModal.title')} size="large">
      <div className="tc__poolPositionModal-info">
        <PoolPositionCard
          apr={apr}
          term={term}
          profitLoss={profitLoss}
          balance={balance}
          totalYieldEarned={totalYieldEarned}
          projectedTotalYield={projectedTotalYield}
          tokenDecimals={tokenDecimals}
          tokenExchangeRate={tokenExchangeRate}
          tokenTicker={tokenTicker}
        />
      </div>
      <div className="tc__poolPositionModal-chart">
        <DateChart data={chartData} dot={chartDot} />
        {/* TODO - Add extra x-axis in chart component that will show 'Earned/Projected yield' labels */}
      </div>
      {/* TODO - Add transaction history (check with design team if it's done) */}
      <div className="tc__poolPositionModal-actions">
        <ActionButton
          onClick={onWithdraw}
          variant="primary"
          size="large"
          labels={{
            default: t('PoolPositionModal.withdrawAction'),
          }}
        />
      </div>
    </Modal>
  );
};
export default PoolPositionModal;
