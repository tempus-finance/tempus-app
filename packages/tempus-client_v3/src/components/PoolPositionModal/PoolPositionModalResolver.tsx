import { FC } from 'react';
import { useParams } from 'react-router-dom';
import { Chain, Decimal, ProtocolName, Ticker } from 'tempus-core-services';
import { usePoolBalance } from '../../hooks';
import { PoolPositionModal } from './PoolPositionModal';

export const PoolPositionModalResolver: FC = () => {
  const { chain, ticker, protocol, poolAddress } = useParams();

  const poolBalanceData = usePoolBalance(poolAddress, chain as Chain);

  // TODO - Properly check if URL params have valid values - if not show an error page or redirect to root page

  return (
    // TODO - Replace dummy data with data from hooks
    <PoolPositionModal
      apr={0.1}
      balance={poolBalanceData?.balance || ZERO}
      chartData={[
        { x: new Date(2022, 3, 4), y: 10 },
        { x: new Date(2022, 4, 1), y: 20 },
      ]}
      profitLoss={new Decimal(2)}
      projectedTotalYield={new Decimal(3)}
      term={new Date(2022, 4, 2)}
      tokenDecimals={4}
      tokenExchangeRate={new Decimal(100)}
      totalYieldEarned={new Decimal(2)}
      tokenTicker="ETH"
      chain={chain as Chain}
      address={poolAddress || ''}
      backingToken={ticker as Ticker}
      protocol={protocol as ProtocolName}
    />
  );
};
