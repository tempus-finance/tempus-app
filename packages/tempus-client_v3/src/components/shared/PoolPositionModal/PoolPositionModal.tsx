import { FC, useCallback } from 'react';
import { Decimal, Ticker } from 'tempus-core-services';
import { ChartDataPoint, ChartDot, DateChart, ChartDotVariant } from '../Chart';
import ActionButton from '../ActionButton';
import Modal from '../Modal';
import PoolPositionCard from '../PoolPositionCard';

import './PoolPositionModal.scss';

interface PoolPositionModalProps {
  apr: number;
  term: Date;
  profitLoss: Decimal;
  balance: Decimal;
  totalYieldEarned: Decimal;
  projectedTotalYield: Decimal;
  tokenDecimals: number;
  tokenExchangeRate: Decimal;
  tokenTicker: Ticker;
  chartData: ChartDataPoint<Date, number>[];
  onWithdraw: () => void;
}

const PoolPositionModal: FC<PoolPositionModalProps> = props => {
  const {
    apr,
    term,
    profitLoss,
    balance,
    totalYieldEarned,
    projectedTotalYield,
    tokenDecimals,
    tokenExchangeRate,
    tokenTicker,
    chartData,
    onWithdraw,
  } = props;

  const chartDot = useCallback(
    (_x, _y, index, cx, cy) => {
      let variant: ChartDotVariant = 'plus';
      if (index > 1) {
        const currentValue = chartData[index].y;
        const previousValue = chartData[index - 1].y;

        if (currentValue < previousValue) {
          variant = 'minus';
        }
      }

      return <ChartDot variant={variant} centerX={cx} centerY={cy} key={`chart-dot-${cx}-${cy}`} />;
    },
    [chartData],
  );

  return (
    <Modal open onClose={() => {}} title="Position Details" size="large">
      <div className="tc__poolPositionModal-info">
        <PoolPositionCard
          apr={apr}
          term={term}
          profitLoss={profitLoss}
          balance={balance}
          totalYieldEarned={totalYieldEarned}
          projectedTotalYield={projectedTotalYield}
          tokenDecimals={tokenDecimals}
          tokenExchangeRate={tokenExchangeRate}
          tokenTicker={tokenTicker}
        />
      </div>
      <div className="tc__poolPositionModal-chart">
        <DateChart data={chartData} dot={chartDot} />
        {/* TODO - Add extra x-axis in chart component that will show 'Earned/Projected yield' labels */}
      </div>
      {/* TODO - Add transaction history (check with design team if it's done) */}
      <div className="tc__poolPositionModal-actions">
        <ActionButton
          onClick={onWithdraw}
          variant="primary"
          size="large"
          labels={{
            default: 'Make Withdrawal',
          }}
        />
      </div>
    </Modal>
  );
};
export default PoolPositionModal;
