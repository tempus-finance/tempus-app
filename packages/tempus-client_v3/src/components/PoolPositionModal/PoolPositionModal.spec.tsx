import { render } from '@testing-library/react';
import { Decimal } from 'tempus-core-services';
import { PoolPositionModal } from './PoolPositionModal';

const subject = () =>
  render(
    <PoolPositionModal
      apr={0.1}
      balance={new Decimal(20)}
      chartData={[
        { x: new Date(2022, 3, 4), y: 10 },
        { x: new Date(2022, 4, 1), y: 20 },
      ]}
      profitLoss={new Decimal(2)}
      projectedTotalYield={new Decimal(3)}
      term={new Date(2022, 4, 2)}
      tokenDecimals={4}
      tokenExchangeRate={new Decimal(100)}
      totalYieldEarned={new Decimal(2)}
      tokenTicker="ETH"
      protocol="lido"
      address="0x123123123"
      backingToken="ETH"
      chain="ethereum"
    />,
  );

describe('PoolPositionModal', () => {
  it('renders a pool position modal', () => {
    const { container } = subject();

    expect(container).not.toBeNull();
    expect(container).toMatchSnapshot();
  });
});
