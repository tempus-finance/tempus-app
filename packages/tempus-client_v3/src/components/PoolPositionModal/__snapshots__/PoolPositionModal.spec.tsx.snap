// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PoolPositionModal renders a pool position modal 1`] = `
<div>
  <div
    class="tc__modal-backdrop"
    id="modal-backdrop"
  />
  <div
    class="tc__modal-container "
  >
    <div
      class="tc__modal tc__modal__plain tc__modal__size-large"
    >
      <div
        class="tc__modal-header"
      >
        <button
          class="tc__btn "
          data-selected="false"
          type="button"
        >
          <svg
            class="tc__icon tc__icon-close"
            fill="none"
            height="20"
            viewBox="0 0 16 16"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.6081 13.7161C15.859 13.967 16 14.3073 16 14.6621C16 15.017 15.859 15.3572 15.6081 15.6081C15.3572 15.859 15.017 16 14.6621 16C14.3073 16 13.967 15.859 13.7161 15.6081L8 9.87872L2.2839 15.6081C2.033 15.859 1.6927 16 1.33788 16C0.98305 16 0.642756 15.859 0.391855 15.6081C0.140955 15.3572 0 15.017 0 14.6621C0 14.3073 0.140955 13.967 0.391855 13.7161L6.12128 8L0.391855 2.2839C0.140955 2.033 -2.64367e-09 1.6927 0 1.33788C2.64367e-09 0.98305 0.140955 0.642756 0.391855 0.391855C0.642756 0.140955 0.98305 2.64367e-09 1.33788 0C1.6927 -2.64367e-09 2.033 0.140955 2.2839 0.391855L8 6.12128L13.7161 0.391855C13.967 0.140955 14.3073 0 14.6621 0C15.017 0 15.3572 0.140955 15.6081 0.391855C15.859 0.642756 16 0.98305 16 1.33788C16 1.6927 15.859 2.033 15.6081 2.2839L9.87872 8L15.6081 13.7161Z"
              fill="#222222"
            />
          </svg>
        </button>
      </div>
      <div
        class="tc__modal__body"
      >
        <div
          style="font-style: normal; font-size: 20px; line-height: 28px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Position Details
        </div>
        <div
          class="tc__poolPositionModal-info"
        >
          <div
            class="tc__poolPositionCard"
          >
            <div
              class="tc__poolPositionCard-row"
            >
              <div
                class="tc__poolPositionCard-cell"
              >
                <div
                  class="tc__poolPositionCard-cellData"
                >
                  <div
                    style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(71, 67, 80); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
                  >
                    APR
                  </div>
                  <div
                    style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'Azeret Mono', monospace; opacity: 1;"
                  >
                    10.00%
                  </div>
                </div>
              </div>
              <div
                class="tc__poolPositionCard-cell"
              >
                <div
                  class="tc__poolPositionCard-cellData"
                >
                  <div
                    style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(71, 67, 80); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
                  >
                    Term
                  </div>
                  <span
                    class="tc__formatted-date"
                  >
                    <div
                      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'Azeret Mono', monospace; opacity: 1;"
                    >
                      02
                    </div>
                    <span
                      class="tc__formatted-date__separator tc__formatted-date__separator-high-contrast"
                    />
                    <div
                      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'Azeret Mono', monospace; opacity: 1;"
                    >
                      May
                    </div>
                    <span
                      class="tc__formatted-date__separator tc__formatted-date__separator-high-contrast"
                    />
                    <div
                      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'Azeret Mono', monospace; opacity: 1;"
                    >
                      24
                    </div>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="tc__poolPositionCard-row"
            >
              <div
                class="tc__poolPositionCard-cell"
                data-size="small"
              >
                <button
                  class="tc__btn "
                  data-selected="false"
                  type="button"
                >
                  <div
                    class="tc__poolPositionCard-seeMore"
                  >
                    <div
                      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
                    >
                      See more
                    </div>
                    Â  
                    <svg
                      class="tc__icon tc__icon-down-chevron"
                      fill="none"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <g
                        clip-path="url(#clip0_909_5439)"
                      >
                        <path
                          d="M15.7387 5.36992L8.63082 11.7636C8.54823 11.8385 8.44996 11.898 8.34169 11.9385C8.23342 11.9791 8.11729 12 8 12C7.88271 12 7.76658 11.9791 7.65831 11.9385C7.55004 11.898 7.45177 11.8385 7.36918 11.7636L0.261296 5.36992C0.0939908 5.21942 -2.49304e-09 5.01531 0 4.80248C2.49304e-09 4.58965 0.0939908 4.38554 0.261296 4.23504C0.428601 4.08455 0.655515 4 0.89212 4C1.12873 4 1.35564 4.08455 1.52294 4.23504L8 10.0693L14.4771 4.23504C14.6444 4.08455 14.8713 4 15.1079 4C15.3445 4 15.5714 4.08455 15.7387 4.23504C15.906 4.38554 16 4.58965 16 4.80248C16 5.01531 15.906 5.21942 15.7387 5.36992Z"
                          fill="#222222"
                        />
                      </g>
                      <defs>
                        <clippath
                          id="clip0_909_5439"
                        >
                          <rect
                            fill="white"
                            height="16"
                            width="16"
                          />
                        </clippath>
                      </defs>
                    </svg>
                  </div>
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="tc__poolPositionModal-chart"
        >
          <div
            class="recharts-responsive-container"
            style="width: 100%; height: 345px;"
          />
        </div>
        <div
          class="tc__poolPositionModal-actions"
        >
          <button
            class="tc__btn tc__actionButton tc__actionButton-border-primary-large tc__actionButton-background-primary-large tc__actionButton-hover-primary-large tc__actionButton-shadow-default tc__actionButton-cursor-default tc__actionButton-size-large "
            data-selected="false"
            type="button"
          >
            <div
              class="tc__actionButton-status-label"
            >
              <div
                style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(255, 255, 255); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
              >
                Make Withdrawal
              </div>
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
